File: G-for-g-Practice/Decode the string.py

class Solution:
    def decodedString(self, s):
        # code here
        Stack =[]
        curr_num =0
        curr_str=""
        
        for char in s:
            if char.isdigit():
                curr_num= curr_num*10 +int(char)
            elif char=='[':
                Stack.append((curr_str,curr_num))
                curr_str,curr_num= "",0
            elif char==']':
                prev_str, num = Stack.pop()
                curr_str = prev_str+ num*curr_str
            else:
                curr_str+=char
        return curr_str

File: G-for-g-Practice/Decode the string.java



class Solution {
    static String decodeString(String s) {
        // code here
        Stack<String> strStack = new Stack<>();
        Stack<Integer> numStack = new Stack<>();
        String currStr ="";
        int currNum=0;
        
        for (char c: s.toCharArray()){
            if(Character.isDigit(c)){
                currNum= currNum * 10 + (c - '0');
            }
            else if (c=='['){
                numStack.push(currNum);
                strStack.push(currStr);
                currStr ="";
                currNum=0;
            }
            else if (c==']'){
                int num = numStack.pop();
                StringBuilder temp = new StringBuilder (strStack.pop());
                temp.append(currStr.repeat(num));
                currStr = temp.toString();
            }
            else{
                currStr+=c;
            }
        }
        return currStr;
    }
}
